quill-core/src/main/scala/io/getquill/norm/AdHocReduction.scala

```diff
@@ -52,6 +44,26 @@ object AdHocReduction {
       case FlatMap(a, b, UnionAll(c, d)) =>
         Some(UnionAll(FlatMap(a, b, c), FlatMap(a, b, d)))
 
+      // query(...).take(n).map(d => e) =>
```

Comment:
I think https://github.com/mentegy/quill/blob/master/quill-core/src/main/scala/io/getquill/norm/OrderTerms.scala would be a better place to add these transformations